### PR TITLE
Add admin panel for bulk ownership changes

### PR DESCRIPTION
**Pull request by joelverhagen**
Spec: https://github.com/NuGet/Engineering/pull/3998
Issue: https://github.com/NuGet/Engineering/issues/3994

This admin panel allows site admins to make bulk ownership changes (add or remove owners) and set an arbitrary user as the "sender" for ownership requests. As site admins we'll use the `NuGetAdmin` user account for this so our own personal name can be kept private.

This is what it looks like:
image

And this is the "validate" screen since the operations are so complex:
image

And this is the admin home page (I've tightened things up a bit since I felt like the display was too spacious)
image

### CODE REVIEW

File: src/NuGetGallery.Services/Configuration/IAppConfiguration.cs [hunk 1]
```diff
@@ -501,5 +501,14 @@ public interface IAppConfiguration : IMessageServiceConfiguration
         /// https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/7w2sway1(v=vs.100)?redirectedfrom=MSDN
         /// </summary>
         int? MaxIoThreads { get; set; }
+
+        /// <summary>
+        /// The username of the user that can be entered as the sender for admin flows. When an admin flow may send one
+        /// or more emails to end users, it helps to mask the identity of the site admin that performed the action by
+        /// using this user instead. This account is not created automatically. The username should refer to a user
+        /// account (not organization) that has an email address that can be visible for administrative notices. This
+        /// account should not have any credentials or be marked as a site admin.
```
Comment (agr) [inline src/NuGetGallery.Services/Configuration/IAppConfiguration.cs:510]: Is the condition from the last sentence actually checked?

**Review by shishirx34 — APPROVED** [review]

File: src/NuGetGallery/Areas/Admin/Controllers/HomeController.cs [hunk 2]
```diff
@@ -24,6 +24,7 @@ public virtual ActionResult Index()
         {
             var viewModel = new HomeViewModel(
                 showDatabaseAdmin: _config.Current.AdminPanelDatabaseAccessEnabled,
+                showLuceneAdmin: _config.Current.SearchServiceUriPrimary == null && _config.Current.SearchServiceUriSecondary == null,
```
Comment (shishirx34) [inline src/NuGetGallery/Areas/Admin/Controllers/HomeController.cs:27]: why is this needed? is it just for on-prem gallery consistency?

File: src/NuGetGallery.Services/Configuration/IAppConfiguration.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGetGallery.Services/Configuration/IAppConfiguration.cs:510]: Nope, thus the "should". It would be pretty pointless to check this at runtime since we have NuGetAdmin set and it will never need to be looked at again.

File: src/NuGetGallery/Areas/Admin/Controllers/HomeController.cs (hunk 2 above)
Comment (joelverhagen) [reply src/NuGetGallery/Areas/Admin/Controllers/HomeController.cs:27]: Yes. This lucene admin page is useless in DEV/INT/PROD so I'd like to hide it.

File: src/NuGetGallery/Areas/Admin/ViewModels/PackageOwnershipState.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery.Areas.Admin.ViewModels
+{
+    /// <summary>
+    /// The possible ways that a user have have ownership change with respect to a package registration. These enum
```
Comment (joelverhagen) [inline src/NuGetGallery/Areas/Admin/ViewModels/PackageOwnershipState.cs:7]:
```suggestion
    /// The possible ways that a user can have ownership change with respect to a package registration. These enum
```

File: src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs [hunk 4]
```diff
@@ -272,7 +282,7 @@ public async Task RemovePackageOwnerAsync(PackageRegistration packageRegistratio
                 throw new ArgumentNullException(nameof(ownerToBeRemoved));
             }
 
-            if (OwnerHasPermissionsToRemove(requestingOwner, ownerToBeRemoved, packageRegistration))
+            if (!requireNamespaceOwnership || OwnerHasPermissionsToRemoveFromNamespace(requestingOwner, ownerToBeRemoved, packageRegistration))
```
[the remark this answers is not in the document]
Comment (joelverhagen) [reply src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs:285]: In the admin for the `requestingOwner` might be `NuGetAdmin` which has no priviledges at all. It is not marked as a site admin because it shouldn't be signed in to directly.

File: src/NuGetGallery/Areas/Admin/ViewModels/PackageOwnershipChangesModel.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            AddOwners = addOwners;
+            RemoveOwners = removeOwners;
+            Changes = changes;
+        }
+
+        public PackageOwnershipChangesInput Input { get; }
+        public User Requestor { get; }
+        public string RequestorUsername => Requestor.Username;
+        public IEnumerable<string> PackageIds => Changes.Select(x => x.Id);
+        public IReadOnlyList<string> AddOwners { get; }
```
Comment (agr) [inline src/NuGetGallery/Areas/Admin/ViewModels/PackageOwnershipChangesModel.cs:30]: Why not `IReadOnlyCollection`?

**Review by shishirx34 — APPROVED** [review]

File: src/NuGetGallery/Areas/Admin/Controllers/PackageOwnershipController.cs [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+        private bool ValidateInput(PackageOwnershipChangesInput input, out ViewResult errorViewResult, out PackageOwnershipChangesModel validatedModel)
+        {
+            errorViewResult = null;
+            validatedModel = null;
+
+            // Find all of the users, by username
+            var requestorUsername = (input.Requestor ?? string.Empty).Trim();
+            if (string.IsNullOrEmpty(requestorUsername))
+            {
+                errorViewResult = ShowError(input, nameof(PackageOwnershipChangesInput.Requestor), "You must provide a requestor username.");
```
[the remark this answers is not in the document]
Comment (joelverhagen) [reply src/NuGetGallery/Areas/Admin/Controllers/PackageOwnershipController.cs:141]: I chose to take the simpler approach since it's an admin tool. Since the input is returned back to the error page, it's relatively painless to work through any error since the previous input is put back into the form.

File: src/NuGetGallery/Areas/Admin/ViewModels/PackageOwnershipChangesModel.cs (hunk 5 above)
Comment (joelverhagen) [reply src/NuGetGallery/Areas/Admin/ViewModels/PackageOwnershipChangesModel.cs:30]: I like to explicitly keep the order (with a list) in a view model so that the controller can manage this if desired. I could store the list in the collection property, but this makes it clear in the view that order is intended.
